crates/flux-middle/src/rty/projections.rs

```diff
@@ -136,7 +137,47 @@ impl<'genv, 'tcx, 'cx> Normalizer<'genv, 'tcx, 'cx> {
         Ok((true, ty))
     }
 
-    fn confirm_candidate(&self, candidate: Candidate, obligation: &AliasTy) -> QueryResult<Ty> {
+    // See issue-829.rs for an example of what this function is for.
+    fn resolve_projection_predicates(
+        &mut self,
+        subst: &mut TVarSubst,
+        impl_def_id: DefId,
+    ) -> QueryResult {
+        // 1. make a partial substitution with the "known" generic arguments e.g.
+        //   Given [T1 ->?, T2 -> i32] make a substitution [T1 -> T1, T2 -> i32]
+        let known_args = subst.as_generic_args(&self.genv, self.genv.generics_of(impl_def_id)?)?;
+
+        // 2. instantiate the generic_predicates of `impl_def_id` with the known_args
+        // e.g. given a predicate `T2 : Trait1<Assoc=T1>` and known_args [T1 -> T1, T2 -> i32]
+        //      we get `i32 : Trait1<Assoc=T1>`
+        let projection_preds: FxHashSet<_> = self
+            .genv
+            .predicates_of(impl_def_id)?
+            .map(|generic_predicates| generic_predicates.predicates)
+            .instantiate(self.tcx(), &known_args, &[])
+            .iter()
+            .filter_map(|pred| {
+                if let ClauseKind::Projection(pred) = pred.kind_skipping_binder() {
+                    Some(pred.clone())
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // 3. recursively solve each of the projection predicates and unify result with the predicate's term
+        // e.g. given `i32 : Trait1<Assoc=T1>` we
+        //      - `normalize_projection_ty` on `<i32 as Trait1>::Assoc` to get `bool`
+        //      - unify `T1` with `bool`
+        for p in projection_preds.iter() {
```

Comment:
I think this is wrong for something like:

```rust
trait Trait1 {
    type Assoc1;
}

impl Trait1 for i32 {
    type Assoc1 = bool;
}

trait Trait2 {
    type Assoc2;
}

struct S<T> {
    fld: T,
}

impl<T1, T2, T3> Trait2 for S<T1>
where
    T2: Trait1<Assoc1 = T3>,
    T1: Trait1<Assoc1 = T2>,
{
    type Assoc2 = T1;
}
```
In this example we first need to process `T1: Trait1<Assoc1 = T2>` (with the solution for `T1`) to solve `T2` before processing `T2: Trait1<Assoc1 = T3>`.

More specifically, we can only process a projection predicate `T1::Assoc = T2` if `T1` is *fully resolved* (note that normalizing something that still has generic parameters is automatically wrong because the parameter is in the context of the implementation, but we normalize in the context of the caller, i.e., the parameter will refer to the wrong thing).

I think we need a loop like so

```
while !pending.is_empty() {
    let ready = extract_fully_resolved(subst, pending);
    for clause in ready {
        // do the thing
    }
}
```

Reply:
I guess this is exactly why applying a partial instantiation is not such a good idea. Doing `[T1 -> T1, T2 -> i32]` is technically instantiating `T1` in the context of the implementation with `T1` in the context of the caller, so one has to be very careful not to mix them.

Reply:
actually, the substitution can genuinely instantiate one of the arguments with a parameter in the caller, so after you apply it is impossible to tell which parameters haven't been resolved and which ones are genuinely solved to a parameter in the caller.

Reply:
totally not following what this caller business is about?

Reply:
Ok I think I know what you mean but the _real_ issue is this: Given

* a `TVarSubst`
* a `EarlyBinder<Ty>`

How do you tell if the `TVarSubst` "resolves" the Ty?

Reply:
I suppose one can walk over the `Ty` to gather the indices (debruin vars) and check if all of those are defined in `TVarSubst`?

Reply:
Or! make the `instantiate` return an `Err` if that parameter is not defined...

Reply:
Either that or to have a fallible instantiation that given a partial substitution returns `None` if the type contains parameters not in the substitution. Not much difference between the two.

(these are early params though so no debruin, but a `ParamTy`)

Reply:
Perhaps we need a separate `instantiate` that takes a "partial" subst, and returns an option/result...

Reply:
A visitor that returns the set of `ParamTy`s feels useful on its own, so I'd go with that.

Reply:
In this case we need to do the actual substitution though (to get the partial instantiation that can then be recursively normalized...) so think the "fallible instantiation" is the way to go?

Reply:
you can implement the fallible instantiation by first checking if all mentioned parameters are solved and then substituting, so it's a matter of whether to check first and then substitute, or check as you substitute. Not much difference. I don't feel strong one way or the other.

Reply:
hmm. the former seems a bit redundant imo, so went with the latter.